Add support for internal enums via stub generation

bwoebi opened the pull request:
This uses persistently malloc()'ed objects, which get duplicated to an request-unique object.
Those request-unique objects are supposed to be immutable for the lifetime of the request, i.e. they can be just efree()'d without issues.

~~This patch requires https://github.com/nikic/PHP-Parser/pull/792 (and a subsequent update of the tag) before merging.~~ Done.

nikic commented on Zend/zend_enum.c:
```diff
@@ -16,11 +16,13 @@
    +----------------------------------------------------------------------+
 */
 
+#define GENERATE_ZEND_ENUM_REGISTRATIONS
```
What's this for?

bwoebi replied:
Avoiding messages in every file including the `.h` about unused static functions (the registration functions) while compiling - this #if enables them.

nikic replied:
Oh, you're including the zend_enum_arginfo.h from the header. Why do you do that? The arginfo.h files are not supposed to be included in headers.

bwoebi replied:
To have the arginfo for tryfrom/from/cases available - every enum needs them.

bwoebi replied:
Also see `ZEND_ENUM_ME_END` and `ZEND_ENUM_BACKED_ME_END`

bwoebi replied:
Actually, I tried to expose them via an extern first - but if I do that via initializer assignment then the compiler complains that I cannot use non-compile time constants in initializers...

nikic replied:
Would it be possible to add the methods in the registration function rather than requiring them to be explicitly listed? Same as userland that is.

bwoebi replied:
I would have to malloc() some array so that the builtin_functions pointer would be adequately populated ... and then free() it ... when the class entry is destroyed. I feel quite unsure about that.
This is definitely more obvious in userland.

bwoebi replied:
Oh, you probably mean directly adding it to the function_table when you say same as userland. that probably could work out...

bwoebi replied:
Actually, I just could call zend_register_functions() - was way easier than I expected.

iluuu1994 reviewed (commented):
Could use more tests, especially things like `unserialize`. Can't comment on persistent objects.

bwoebi commented on build/gen_stub.php:
```diff
@@ -995,6 +1023,100 @@ private function appendMethodSynopsisTypeToElement(DOMDocument $doc, DOMElement
+
+                case "string":
+                    $suffix = "_string";
+                    $additionalParams = ", \"$value\"";
```
Yes, I've been inspired by other places using string values… but an addslashes() would do no harm though

bwoebi commented on Zend/zend_enum.h:
```diff
@@ -47,6 +48,22 @@ static zend_always_inline zval *zend_enum_fetch_case_value(zend_object *zobj)
 	return OBJ_PROP_NUM(zobj, 1);
 }
 
+ZEND_API zend_class_entry *zend_register_internal_enum(const char *name);
```
Depends … well, you _have to know_ that IS_UNDEF means unbacked and NULL means no functions (i.e. you don't need to define an empty array with just ZEND_FE_END). It's not always obvious what you would have to do if these simpler functions wouldn't be there.

Especially the distinction backed vs. unbacked makes sense to me

bwoebi commented:
Good point, didn't think about deserialization at all. Will check that out. Also need to add a test for assignment to simple constant (there shall be no constant propagation of enum objects).

bwoebi commented:
Deserialization indeed had only ZEND_COPY instead of ZEND_COPY_OR_DUP. Fixed that. Added test. Anything else you can think of where a test would be needed? @iluuu1994 
Note that for most cases it should be identical to userland enums, just that the persistent ZEND_OBJECT should never leak into userland. I.e. all places possibly accessing class constants must prevent their propagation, but that's in general covered by ZEND_COPY_OR_DUP() which applies also to internal arrays being in class constants.

nikic commented on Zend/zend_objects.c:
```diff
@@ -282,3 +289,27 @@ ZEND_API zend_object *zend_objects_clone_obj(zend_object *old_object)
+	_zend_object_minimal_init(obj, ce);
+	GC_TYPE_INFO(obj) |= IS_OBJ_PERSISTENT << GC_FLAGS_SHIFT;
+	obj->handle = zend_objects_persistent_handle_counter;
+	zend_objects_persistent_handle_counter += (alloc_size - 1) / sizeof(zend_object) + 1;
```
If you're doing something like this anyway, why is persistent_objects/zend_objects_persistent_handle_counter quantized to zend_objects rather than just a raw memory region?

bwoebi replied:
Probably does not make too much sense, the main reason is that I'm accessing `EG(persistent_objects) + obj->handle` where EG(persistent_objects) is a zend_object-array. But could just cast to char and back to zend_object at that point.

nikic replied:
I mean, why is it `zend_object*` in the first place, rather than `char*`?

bwoebi replied:
Because it contains objects :-D that was the logical thought I had, but obviously, this chunk of memory can just be made more generic and the zend_objects_persistent_handle_counter just be a counter. There's no reason why this chunk of memory cannot be just shared between multiple potential usages. (and then I'd also decouple that "handle_counter" from zend_objects.c)

nikic reviewed (commented):
Okay, so here is my understanding of this patch:
 * Persistent objects are `malloc`ed and get a `GC_PERSISTENT` flag. We don't have space for this flag, so one bit of the GC root space is stolen.
 * Persistent objects cannot be returned into userland code and need to be duplicated first. It looks like you missed doing this at least in the from/tryFrom and cases implementation. Try running with ZEND_RC_DEBUG=1: https://gist.github.com/nikic/5bd0c0f1280c217c957ef8fa55c82f8e.
 * Persistent objects are duplicated in the following way: There is a single allocation for all persistent objects, whose number and size is known at startup. The persistent object stores the offset into this allocation. If the refcount of the corresponding object is zero, it is initialized once using a shallow copy. All per-request objects are freed at once at the end of the request and cannot be individually freed.

I think the principal alternative to this approach is to do the same as in userland, i.e. use CONSTANT_AST constant values that evaluate to the (non-persistent objects). However, this would require support for mutable_data for internal classes. I think overall this may be the cleaner solution, because it brings handling of internal and userland classes closer and does not require introduction of any new concepts.

Anyway, I would like to hear @dstogov's thoughts on this patch and the persistent object concept.

bwoebi commented on Zend/zend_enum.c:
```diff
@@ -371,3 +376,128 @@ void zend_enum_register_props(zend_class_entry *ce)
+	return zend_register_internal_backed_enum_ex(name, IS_UNDEF, NULL);
+}
+
+ZEND_API zend_class_entry *zend_register_internal_enum_ex(const char *name, const zend_function_entry *functions)
```
Yeah I tend to agree, I should just distinguish backed vs unbacked.

bwoebi commented on Zend/zend_enum.c:
```diff
@@ -371,3 +376,128 @@ void zend_enum_register_props(zend_class_entry *ce)
+	zend_class_constant *c = zend_declare_class_constant_ex(ce, name, &casezv, ZEND_ACC_PUBLIC, NULL);
+	ZEND_CLASS_CONST_FLAGS(c) |= ZEND_CLASS_CONST_IS_CASE;
+
+	zend_string_release(name);
```
I feel it's unergonomic to do `zend_string * ... = zend_string_init(..); zend_register_enum_case(...); zend_string_release(...);` it essentially becomes three statements instead of just one `zend_register_enum_case(... zend_string_init (..))` … I particular I cannot imagine a case where you would want to retain the zend_string afterwards.

bwoebi commented:
> It looks like you missed doing this at least in the from/tryFrom and cases implementation.

Yes I did. Arghs. I checked around where constants are accessed, but forgot to check the parts where the enums are directly accessed :-/

> I think the principal alternative to this approach is to do the same as in userland, i.e. use CONSTANT_AST constant values that evaluate to the (non-persistent objects). However, this would require support for mutable_data for internal classes. I think overall this may be the cleaner solution, because it brings handling of internal and userland classes closer and does not require introduction of any new concepts.

You precisely found the issue. Doing the same as userland is not so easy though, because the constant gets replaced by the unique object. (in userland) ... In internal there is just one overall copy of the constants table. If we'd replace it there with ast evaluation, it would break. Creating a new object obviously is out of question as well.

Also, how would you even implement that mutable data? In particular with ZTS, you need thread local mutable data. Which presumably would live somewhere in EG() in some mapping I guess. At least it's not possible to just do ce->thread_local_mutable_data = &zend_tls_marked_variable; because the address is unique per thread... 
At that point, obviously, we could have an AST node, which, when evaluated goes to look up into the mapping for mutable class data and from there - needs to initialize that mutable data if not present yet, then check if the object exists and if not, do a shallow copy as emalloc()'ed refcountable object.

Maybe I'm overestimating what would need to be done, but yeah, how do you imagine that working out? _If_ you find a nice easy way.

TysonAndre commented on Zend/zend_enum.c:
```diff
@@ -371,3 +376,128 @@ void zend_enum_register_props(zend_class_entry *ce)
+	ZVAL_STR(&zv, str);
+	zend_object *case_obj = add_internal_enum_case(ce, name, &zv);
+
+	zend_string_release(name);
+	zend_string_release(str);
```
```c
static zend_always_inline void zend_string_release(zend_string *s)
{
	if (!ZSTR_IS_INTERNED(s)) {
		if (GC_DELREF(s) == 0) {
			pefree(s, GC_FLAGS(s) & IS_STR_PERSISTENT);
		}
	}
}
```

```suggestion
```

TysonAndre commented on Zend/zend_enum.c:
```diff
@@ -371,3 +376,128 @@ void zend_enum_register_props(zend_class_entry *ce)
+	str = zend_new_interned_string(str);
+
+	zval namezv;
+	ZVAL_STR_COPY(&namezv, name);
```
```suggestion
	ZVAL_INTERNED_STR(&namezv, name);
```

TysonAndre commented on Zend/zend_enum.c:
```diff
@@ -371,3 +376,128 @@ void zend_enum_register_props(zend_class_entry *ce)
+	}
+
+	zval zv;
+	ZVAL_STR(&zv, str);
```
```suggestion
	ZVAL_INTERNED_STR(&zv, str);
```

TysonAndre commented on Zend/zend_enum.c:
```diff
@@ -371,3 +376,128 @@ void zend_enum_register_props(zend_class_entry *ce)
+	ZVAL_LONG(&zv, value);
+	zend_object *case_obj = add_internal_enum_case(ce, name, &zv);
+
+	zend_string_release(name);
```
Same comments on ZVAL_INTERNED_STR and not needing zval_string_released on zend_new_interned_string. It seems like both zend_string_init_interned_permanent and zend_new_interned_string_request return interned strings where refcounting or freeing isn't necessary.

The code works but it's longer than it needs to be?

nikic commented:
I've posted an alternative implementation at #7302.

nikic commented:
Added the stub generation part in https://github.com/php/php-src/commit/32d48212ea365803c6dbe02360fb03c9fcfeefb3, so internal enum support should now be complete.
